classes/migration/upgrade/v3_4_0/I5716_EmailTemplateAssignments.php

```diff
@@ -0,0 +1,505 @@
+                ->get();
+
+            collect([
+                'EDITOR_ASSIGN_REVIEW' => 'DISCUSSION_NOTIFICATION_REVIEW',
+                'EDITOR_ASSIGN_PRODUCTION' => 'DISCUSSION_NOTIFICATION_PRODUCTION',
+            ])->each(function(string $alternateTo, string $key) use ($rows, $contextId) {
+
+                DB::table('email_templates')->insert([
+                    'email_key' => $key,
+                    'context_id' => $contextId,
+                    'alternate_to' => $alternateTo,
+                ]);
+
+                $emailId = DB::getPdo()->lastInsertId();
+
+                DB::table('email_templates_settings')
+                    ->insert(
+                        $rows->map(
+                            function(stdClass $row) use ($emailId) {
+                                return [
+                                    'email_id' => $emailId,
+                                    'locale' => $row->locale,
+                                    'setting_name' => $row->setting_name,
+                                    'setting_value' => $row->setting_value,
+                                ];
+                            }
+                        )->toArray()
+                    );
+            });
+        });
+    }
+
+    /**
+     * Reset the editor assign template to its original key
+     */
+    protected function downgradeEditorAssignTemplate(): void
+    {
+        DB::table('email_templates')
+            ->where('email_key', 'EDITOR_ASSIGN_SUBMISSION')
+            ->update(['email_key' => 'EDITOR_ASSIGN']);
+
+        DB::table('email_templates_default_data')
+            ->where('email_key', 'EDITOR_ASSIGN_SUBMISSION')
+            ->update(['email_key' => 'EDITOR_ASSIGN']);
+
+        $emailIds = DB::table('email_templates')
+            ->where('email_key', 'EDITOR_ASSIGN_REVIEW')
+            ->orWhere('email_key', 'EDITOR_ASSIGN_PRODUCTION')
+            ->pluck('email_id');
+
+        DB::table('email_templates')
+            ->whereIn('email_id', $emailIds->toArray())
+            ->delete();
+
+        DB::table('email_templates_settings')
+            ->whereIn('email_id', $emailIds->toArray())
+            ->delete();
+    }
+
+    /**
+     * Assign all remaining unassigned custom templates to the
+     * discussion stage mailables.
+     *
+     * Assigns the existing custom template to the first discussion
+     * mailable that exists. Then creates copies of the template for
+     * each of the remaining discussions.
+     */
+    protected function assignRemainingCustomTemplates(Collection $contextIds): void
+    {
+        $newIds = [];
+        $contextIds->each(function(int $contextId) {
+
+            DB::table('email_templates as et')
+                ->leftJoin('email_templates_default_data as etdd', 'et.email_key', '=', 'etdd.email_key')
+                ->where('et.context_id', $contextId)
+                ->whereNull('et.alternate_to')
+                ->whereNull('etdd.email_key')
```

Comment:
This method shouldn't be touching any of email templates with a default email template key -- whether it has been modified or not. This is only for totally custom emails, which are not accessible in the new UI unless they are assigned as an alternate to a default template.

I think I understand what you're saying about `NOTIFICATION_CENTER_DEFAULT` template. However, `migrateNotificationCenterTemplates()` should only touch modified templates (`email_templates`) and `installTemplate()` should only insert data into the default table (`email_templates_default_data`).

Reply:
I mean that this introduces a bug and forces discussion templates to be assigned as existing custom mailables to to the first default discussion template by this script. 
As far as I remember, during testing, the `email_templates` looked like:
```
`email_key` |  `assignedTo`
DISCUSSION_NOTIFICATION_SUBMISSION | DISCUSSION_NOTIFICATION_SUBMISSION
DISCUSSION_NOTIFICATION_COPYEDITING | DISCUSSION_NOTIFICATION_SUBMISSION
DISCUSSION_NOTIFICATION_PRODUCTION | DISCUSSION_NOTIFICATION_SUBMISSION
```
This is because this line ` ->whereNull('etdd.email_key')`, also retrieves discussion templates added in `migrateNotificationCenterTemplates()` because at the time this line is executed, they have no correspondence in `email_templates_default_data` but installed after, by `installTemplate()`

Reply:
Thanks, I was able to reproduce this in an upgrade from 3.3. I'll look into it.